tdx-tdcall/src/asm/tdcall.asm

```diff
@@ -52,22 +53,25 @@ asm_td_call:
         # tdcall
         .byte 0x66,0x0f,0x01,0xcc
 
-        # Exit if tdcall reports failure.
-        test rax, rax
-        jnz td_call_exit
+        # On TDVM exit, RDI loses the RCX value (TdVmcallArgs struct) that was saved earlier.
+        # But r15 register is not used and is cleared to 0 on TDVM exit so use to it get the
+        # TdcallArgs struct stored on the stack.
+        pop r15
```

Comment:
I am a little confused here. Why not `pop rdi`?

Reply:
As per the ABI spec, `TDG.VP.ENTER` modifies `rdi` to store the `CS base address`. So preferred to use a register which is not used by the call. 
![image](https://github.com/confidential-containers/td-shim/assets/37005992/769e6492-a705-4e8a-986c-55b64cb10d12)

Reply:
That just means RDI is used in TDX module. It does not mean the caller wrapper cannot use it.

See code at line 76, RDI is popped and overridden again.

I agree with you on the old code bug, that RDI is lost and we should not use it directly. But just to pop the saved RCX to RDI is enough. I don't see the need to use R15.

Do you see any failure when you pop RDI?

Reply:
No, I don't see failure when popping to RDI. I have reverted and use RDI. PTAL.